lib/dao.js

```diff
@@ -624,7 +624,193 @@ DataAccessObject.upsert = function(data, options, cb) {
+        where = {};
+      }
+    } else if (cb === undefined) {
+      if (typeof options === 'function') {
+        cb = options;
+        options = {};
+      }
+    }
+    data = data || {};
+    options = options || {};
+    cb = cb || utils.createPromiseCallback();
+    assert(typeof where === 'object', 'The where argument must be an object');
+    assert(typeof data === 'object', 'The data argument must be an object');
+    assert(typeof options === 'object', 'The options argument must be an object');
+    assert(typeof cb === 'function', 'The cb argument must be a function');
+    var hookState = {};
+    var self = this;
+    var Model = this;
+    var connector = Model.getConnector();
+    var modelObject = connector._models;
+    var modelName = Object.keys(modelObject)[0];
+    var id = getIdValue(this, data);
+    var context = {
+      Model: Model,
+      query: { where: where },
+      hookState: hookState,
+      options: options,
+    };
+    Model.notifyObserversOf('access', context, doUpsertWithNPK);
+    function doUpsertWithNPK(err, ctx) {
+      if (err) return cb(err);
+      var isOriginalQuery = isWhereByGivenId(Model, ctx.query.where, id);
+      if (!Object.keys(data).length) {
+        var err = new Error('data object cannot be empty!');
+        err.statusCode = 400;
+        process.nextTick(function() { cb(err); });
+        return cb.promise;
+      }
+      if (connector.upsertWithWhere && !isOriginalQuery) {
+        var context = {
+          Model: Model,
+          where: ctx.query.where,
+          data: data,
+          hookState: hookState,
+          options: options,
+        };
+        Model.notifyObserversOf('before save', context, function(err, ctx) {
+          if (err) return cb(err);
+          data = ctx.data;
+          var update = data;
+          var inst = data;
+          if (!(data instanceof Model)) {
+            inst = new Model(data, { applyDefaultValues: false });
+          }
+          update = inst.toObject(false);
+          Model.applyProperties(update, inst);
+          Model = Model.lookupModel(update);
+          var connector = self.getConnector();
+          var doValidate = undefined;
+          if (options.validate === undefined) {
+            if (Model.settings.automaticValidation !== undefined) {
+              doValidate = Model.settings.automaticValidation;
+            } else {
+              doValidate = Model.settings.validateUpsert;
+            }
+          } else {
+            doValidate = options.validate;
+          }
+          if (doValidate === false) {
+            callConnector();
+          } else {
+            inst.isValid(function(valid) {
+              if (!valid) {
+                if (doValidate) {
+                  return cb(new ValidationError(inst), inst);
+                } else {
+                  console.warn('Ignoring validation errors in updateOrCreate():');
+                  console.warn('  %s', new ValidationError(inst).message);
```

Comment:
Yeah, I agree this is suspicious.

IIRC, the original idea was to *always* run the validation, and depending on `doValidate` flag, either report errors via callback, or log them to console. See how `updateAttributes` is implemented.